File: backend/app/schemas/audit.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Mutlu Erol
"""Audit query API şemaları."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any
from uuid import UUID

from pydantic import BeforeValidator, BaseModel, ConfigDict, Field


def _ip_to_str(v):
    """IPv4Address/IPv6Address objelerini string'e çevir (INET column dönüşümü)."""
    return str(v) if v is not None else None


IPStr = Annotated[str | None, BeforeValidator(_ip_to_str)]


class EventLogPublic(BaseModel):
    """Tek bir audit event."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    occurred_at: datetime
    server_node: str | None = None
    category: str
    event_code: str
    severity: str
    actor_type: str | None = None
    actor_id: str | None = None
    actor_display: str | None = None
    target_type: str | None = None
    target_id: str | None = None
    target_display: str | None = None
    ip_address: IPStr = None
    user_agent: str | None = None
    request_id: IPStr = None
    details: dict[str, Any] = Field(default_factory=dict)


class AuditEventListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[EventLogPublic]


class AuditSummaryBucket(BaseModel):
    bucket: str
    count: int


class AuditSummary(BaseModel):
    """Son N saatlik özet."""

    period_hours: int
    total_events: int
    by_severity: dict[str, int]
    by_category: dict[str, int]
    top_event_codes: list[dict[str, Any]]
    top_actors: list[dict[str, Any]]
    failed_login_count: int
    successful_login_count: int
    timeline: list[AuditSummaryBucket]

File: frontend/src/types/audit.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Mutlu Erol
//
// Backend: app/schemas/audit.py — EventLogPublic, AuditEventListResponse, AuditSummary

export type AuditSeverity =
  | "INFO"
  | "NOTICE"
  | "WARNING"
  | "ERROR"
  | "CRITICAL";

export interface EventLog {
  id: string;
  occurred_at: string;
  server_node: string | null;
  category: string;
  event_code: string;
  severity: string; // AuditSeverity (string olarak gelir)
  actor_type: string | null;
  actor_id: string | null;
  actor_display: string | null;
  target_type: string | null;
  target_id: string | null;
  target_display: string | null;
  ip_address: string | null;
  user_agent: string | null;
  request_id: string | null;
  details: Record<string, unknown>;
}

export interface AuditEventListResponse {
  total: number;
  page: number;
  page_size: number;
  items: EventLog[];
}

export interface AuditSummaryBucket {
  bucket: string;
  count: number;
}

export interface AuditSummary {
  period_hours: number;
  total_events: number;
  by_severity: Record<string, number>;
  by_category: Record<string, number>;
  top_event_codes: Array<{ event_code: string; count: number } & Record<string, unknown>>;
  top_actors: Array<{ actor_id?: string; actor_display?: string; count: number } & Record<string, unknown>>;
  failed_login_count: number;
  successful_login_count: number;
  timeline: AuditSummaryBucket[];
}

export interface AuditEventQuery {
  page?: number;
  page_size?: number;
  category?: string;
  event_code?: string;
  severity?: string;
  actor_id?: string;
  actor_display?: string;
  target_id?: string;
  ip_address?: string;
  server_node?: string;
  search?: string;
  date_from?: string; // ISO
  date_to?: string; // ISO
}
